### PR TITLE
Move model instances into one source file, so they get initialized in the correct order

### PR DESCRIPTION
**Pull request by ChrisDodd**

### CODE REVIEW

**Comment by sethfowler** [comment]
In the long term, the right fix here is to minimize our use of global variables and not depend implicitly on static initialization order.

File: frontends/p4/coreLibrary.cpp [hunk 1]
```diff
@@ -15,9 +15,19 @@ limitations under the License.
...
 
 namespace P4 {
 
 P4CoreLibrary P4CoreLibrary::instance;
 
 }  // namespace P4
+
+/* These must be in the same compiliation unit to ensure that P4CoreLibrary::instance
+ * is initialized before V1Model::instance */
+namespace P4V1 {
```
Comment (mbudiu-vmw) [inline frontends/p4/coreLibrary.cpp]: The name of this cpp file should be changed.
It should perhaps be called "modelInstances.cpp"

**Comment by mbudiu-vmw** [comment]
This was meant to be a singleton pattern, but I guess in C++ you have to do much more work to create one.

**Comment by sethfowler** [comment]
It's not really that much work. =) Both of the approaches Antonin suggested require only a couple of lines of code - for the raw pointer approach, all you need to do is explicitly initialize the variable somewhere before it's used, while for the singleton pattern you just need to wrap accesses to the variable in a function and perform initialization in that function. Call sites need to be updated in both cases, though.

**Comment by ChrisDodd** [comment]
renamed the file to modelInstances.cpp

**Pull request merged**